kartRacer/kart.ts

```diff
@@ -20,40 +36,123 @@ export class Kart extends TransformNode {
 
+    public activateKartCamera(): void {
+        this.setup3rdPersonKartCamera();
+
+        this._scene.registerBeforeRender(() =>
+        {
+            this.beforeRenderUpdate();
+        });
+    }
+
+    private updateFromPhysics(): void {
+        var ray = new Ray(this.position, this.up.scale(-1.0), 0.7);
+        var hit = KartEngine.instance.scene.pickWithRay(ray);
+        if (hit.hit) {
+            var normal = hit.getNormal(true, true);
+
+            // TODO: HACK! It seems the normals of the track are weird, or something, not sure what exactly is going on.
+            normal.scaleInPlace(-1.0)
+
+            this._filteredUp = Vector3.Lerp(
+                this._filteredUp, 
+                normal, 
+                Kart.UP_GROUNDED_FILTER_STRENGTH * this._deltaTime);
+            this._filteredUp.normalize();
+
+            this.position = hit.pickedPoint.add(this._filteredUp.scale(0.55));
+
+            this._velocity.subtractInPlace(normal.scale(Vector3.Dot(this._velocity, normal)));
+
+            this._fallTime = 0.0;
+            this._lastSafePosition.copyFrom(this.position);
+            this._lastSafeFilteredUp.copyFrom(this._filteredUp);
+        }
+        else {
+            this._filteredUp = Vector3.Lerp(
+                this._filteredUp, 
+                Vector3.Up(), 
+                Kart.UP_FALLING_FILTER_STRENGTH * this._deltaTime);
+            this._filteredUp.normalize();
+
+            this._velocity.addInPlace(Vector3.Down().scale(this._deltaTime));
+
+            this._fallTime += this._deltaTime;
+            if (this._fallTime > Kart.MAX_FALL_TIME_SECONDS) {
+                this.position.copyFrom(this._lastSafePosition);
+                this._filteredUp.copyFrom(this._lastSafeFilteredUp);
+                this._velocity.set(0.0, 0.0, 0.0);
+                this._relocity = 0.0;
+            }
+        }
+
+        var forward = Vector3.Cross(this.right, this._filteredUp);
+        var right = Vector3.Cross(this._filteredUp, forward);
+        this.rotationQuaternion = Quaternion.RotationQuaternionFromAxis(right, this._filteredUp, forward);
+    }
+
+    private getForward(): number {
+        //return false ? 1.0 : 0.0;
+        return Math.max(0, Math.min(1, this._input.accelerate));
+    }
+    
+    private getLeft(): number {
+        //return false ? 1.0 : 0.0;
+        return Math.max(0, Math.min(1, -this._input.horizontal));
+    }
+    
+    private getBack(): number {
+        //return false ? 1.0 : 0.0;
+        return Math.max(0, Math.min(1, -this._input.accelerate));
+    }
+    
+    private getRight(): number {
+        //return false ? 1.0 : 0.0;
+        return Math.max(0, Math.min(1, this._input.horizontal));
+    }
+    
+    private getBrake(): number {
+        //return false ? 1.0 : 0.0;
+        return Math.max(0, Math.min(1, this._input.brake));
+    }
+    
+    private updateFromControls(): void {
+        this._turnFactor = Kart.TURN_FILTER_STRENGTH * this.getLeft();
```

Comment:
I think we can combine the Left() and Right() cases to use horizontal

Reply:
I think we can (assuming we persist with input axes), but that's something I'd like to leave to subsequent PRs. This code pretty directly matches what exists in the Unity prototype, so it's not really making any new decisions. Unless you feel strongly otherwise, I'd just prefer to keep this PR as a direct port from Unity so that we have that as part of the history, then build from there.